File: Patterns/Composite Pattern/Cpp/Menu/MenuItem.cpp
#pragma once

#include <iostream>
#include <string>
#include "MenuComponent.cpp"

using namespace std;

class MenuItem : public MenuComponent {
private:
    string name;
    string description;
    bool vegetarian;
    double price;

public: 
    MenuItem(string name, string description, bool vegetatian, double price)
        : name(name), description(description), vegetarian(vegetarian), price(price) {}
    
    string getName() override { 
        return name;
    }

    string getDescription() override {
        return description;
    }

    bool isVegetarian() override {
        return vegetarian;
    }

    double getPrice() override {
        return price;
    }

    void print() override {
        cout << "  " << getName();
        if (isVegetarian()) cout << "(v)";
        cout << ", " << getPrice() << "\n";
        cout << "  -- " << getDescription() << "\n";
    }
};

File: Patterns/Composite Pattern/CSharp/Menu/MenuItem.cs
namespace Menu
{
    public class MenuItem : MenuComponent
    {
        private string? name;
        private string? description;
        private bool isVegetarian;
        private double price;

        public MenuItem(string? name, string? description, bool vegetarian, double price) {
            Name = name;
            Description = description;
            Vegetarian = vegetarian;
            Price = price;
        }

        public override string? Name { get => name ?? "NoText"; set => name = value; }

        public override string? Description { get => description ?? "NoText"; set => description = value; }

        public override bool Vegetarian { get => isVegetarian; set => isVegetarian = value; }

        public override double Price { get => price; set => price = value; }

        public override void Print() {
            Console.Write($"  {name}");
            if (isVegetarian) Console.Write("(v)");
            Console.WriteLine($", {price}");
            Console.WriteLine($"  -- {description}");
        }
    }
}